File: pset2/vigenere/vigenere.c
#include <stdio.h>
#include <math.h>
#include <cs50.h>
#include <ctype.h>
#include <string.h>
#include <stdlib.h>


int main(int argc, string argv[])
{
    // check whether number of command line args = 1 (not including file name)
    if (argc <= 1 || argc >= 3)
    {
        printf("error[1]: Must input only one argument\n");
        return 1;
        void exit(int status);
    }
    else if (true)
    {
        for (int i = 0; i < strlen(argv[1]); i++)
        {
            if (! isalpha(argv[1][i]))
            {
                printf("error[2]: Command line argument must be alphabetical with no spaces.\n");
                return 1;
                break;
                void exit(int status);
            }
        }

        // collecting key from user
        string key = argv[1];
        // get key length
        int keyLength = strlen(key);

        // collecting text to be encrypted
        string text = "";
        text = get_string("plaintext: ");
        // get string length
        int strLength = strlen(text);
        // printf("Length: %i\n", length);

        // obtain key ascii values for string shift
        char letter;
        int keyAsciiCode;
        int keyFormattedAscii;
        int keyNewAsciiCode;
        int keyShifts[100] = {};
        // char newLetter = 'a';
        // char cipher[length + 1];
        printf("ciphertext: ");
        for (int i = 0; i < keyLength; i++)
        {
            letter = key[i];
            // conversion to ascii
            keyAsciiCode = (int) letter;
            // upper or lower or special character (wrap implemented in nested if)
            if (isupper(letter))
            {
                keyFormattedAscii = keyAsciiCode - 65;
            }
            else if (islower(letter))
            {
                keyFormattedAscii = keyAsciiCode - 97;
            }
            else
            {
                keyFormattedAscii = keyAsciiCode;
            }
            // strcat(keyShifts, keyFormattedAscii);
            keyShifts[i] = keyFormattedAscii;
            // printf("%i", keyShifts[i]);
        }

        //perform cipher on input text
        int textAsciiCode;
        int textFormattedAscii;
        int textNewAsciiCode;
        int j = 0;
        for (int i = 0; i < strLength; i++)
        {
            letter = text[i];
            // conversion to ascii
            textAsciiCode = (int) letter;
            // assigning new ascii code after performing shift
            if (isalpha(letter))
            {
                textNewAsciiCode = textAsciiCode + keyShifts[j % keyLength];
                j++;
                //nested wrap implementation
                if (isupper(letter))
                {
                    if (textNewAsciiCode > 'Z')
                    {
                        textNewAsciiCode -= 26;
                    }
                }
                else if (islower(letter))
                {
                    if (textNewAsciiCode > 'z')
                    {
                        textNewAsciiCode -= 26;
                    }
                }

            }
            else
            {
                textNewAsciiCode = textAsciiCode;
            }

            printf("%c", textNewAsciiCode);
        }
        // printf("%i\n", keyShifts);
        printf("\n");
    }

}

File: pset6/vigenere/vigenere.py
"""Implement a program that encrypts messages using Vigenère’s cipher"""

import cs50
import sys



def main():
    """Check for one argument"""
    if len(sys.argv) > 2:
        print("error: Program only accepts one argument")
        sys.exit(1)

    """Check whether argument is alpha"""
    if not sys.argv[1].isalpha():
        print("error: Argument must only contain alphabetical characters")
        sys.exit(1)

    key = sys.argv[1]

    """get string from user"""
    plaintext = cs50.get_string("plaintext: ")

    ALPHALEN = 26
    keyLen = len(key)
    k = []

    """find integer values of key"""
    for i in key:
        if ord(i) >= ord("A") and ord(i) <= ord("Z"):
            k.append(ord(i) - 65)                # k is an int array
        elif ord(i) >= ord("a") and ord(i) <= ord("z"):
            k.append(ord(i) - 97)               # k is an int array
    # print(f"{k}")


    j = 0
    cipher = []

    """perform key additions on cipher"""
    for i, l in enumerate(plaintext):
        if l.isalpha():
            # print(f"{i}")

            """build up cipher string"""
            if plaintext[i].isupper():
                cipher.append(chr((ord(plaintext[i]) + k[j]))) # convert plain text chars to ascii, then add key then convert back to ascii
                if ord(cipher[i]) > 90:            # wrapping implementation
                    cipher[i] = chr(ord(cipher[i]) - 26)

            elif plaintext[i].islower():
                cipher.append(chr((ord(plaintext[i]) + k[j]))) # convert plain text chars to ascii, then add key then convert back to ascii
                if ord(cipher[i]) > 122:            # wrapping implementation
                    cipher[i] = chr(ord(cipher[i]) - 26)
            j += 1

        else:
            """if plain text char isnt alpha"""
            cipher.append(plaintext[i]) # append as is

        """reset j at end of key"""
        if j == keyLen:
            j = 0

    # print(f"{cipher}")

    ciphertext = "".join(cipher)

    print(f"ciphertext: {ciphertext}")



if __name__ == "__main__":
    main()